### ISSUE TITLE
PermissionError: [WinError 32] 

### ISSUE THREAD

**Issue by qq703048949** [Delgan/loguru#202]
from loguru import logger
from multiprocessing import Pool
logger.add('runtime.log', format="{message}", level="INFO", rotation='1MB', retention=5, encoding='utf-8', enqueue=True)
@logger.catch
def my_function(x, y, z):
    return 1 / (x + y + z)
def wrire_log(x):
    for i in range(10000):
        my_function(0, 0, 0)
if __name__ == '__main__':
    p = Pool(10)
    p.map(wrire_log, list(range(10)))
    print('Waiting for all subprocesses done...')
    p.close()
    p.join()
    print('All subprocesses done.')

when I use it ， find  error  for 'PermissionError: [WinError 32] 另一个程序正在使用此文件，进程无法访问。' I need  you help!

**Comment by Delgan** [comment]
@qq703048949 This is most likely a problem with `multiprocessing`: on Windows, your program will be re-executed while spawning a new process, which means `logger.add()` will be called twice, which implies two processes will try to write to the same file and probably caused the error.

Please, carefully read the following documentation which should help you to understand the problem: 
* [Compatibility with `multiprocessing` using `enqueue` argument](https://loguru.readthedocs.io/en/stable/resources/recipes.html#compatibility-with-multiprocessing-using-enqueue-argument)

The workaround is that you should only call `logger.add()` inside your `if __name__ == "__main__"` block. Then, you have to use the `initializer` argument of your `Pool()` object so that each process inherit of the same `logger` with the same handlers.

**Comment by qq703048949** [comment]
If I have to run a py file twice, is there a problem with Linux?

**Comment by Delgan** [comment]
@qq703048949 Linux is more permissive that Windows, but if possible you should avoid writing to the same file from multiple processes without proper synchronization.

**Comment by Delgan** [comment]
@qq703048949 Did you find a solution?

**Comment by Delgan** [comment]
Closing as it looks like file permission error, not a Loguru bug. If you need help to investigate this problem further, just ask me.

**Comment by dushitaoyuan** [comment]
first ,i am a python new learner

I meet the same exception, becase i code like this:
` logger.add("log.log", level=log_level, rotation="500 MB", enqueue=True, encoding='utf-8',
               retention=log_max_days)
    logger.add("log.log", rotation="00:00")`
i debug and find problem  in _file_sink.py at number 205 :
            if new_path == old_path:
                creation_time = get_ctime(old_path)
                root, ext = os.path.splitext(old_path)
                renamed_path = generate_rename_path(root, ext, creation_time)
205:          os.rename(old_path, renamed_path)
                old_path = renamed_path

maybe i set log rotation="00:00",the loguru find the writing log file is old and rename the log file,but the writing log file is using,so the program make exception
may be i should set log_file=log_{time}.log, but that's what I want,
sometimes people want the log file can rotation with max_size and day,just like java logback ,
does loguru support?

**Comment by Delgan** [comment]
Hi @dushitaoyuan.

The error probably comes from the fact that you're adding two sinks both logging to the same file `"log.log"`. Sinks are independent, of each others, so when the first handler will execute the rotation process, it will conflict with the other one which is not aware that a rotation is occurring.

You did well to open a new issue #266, let's keep the discussion over there. :+1:

**Comment by Pixel-Minions** [comment]
Hello,
I am currently facing the same issue, on a render farm. I wonder if you have already tested filelock? I am thinking that if only one process can access the log at the time it can help to avoid this issue.

**Comment by Delgan** [comment]
@Pixel-Minions My guess is that `PermissionError` happens when `logger.add("some_file.log")` is called twice but in different processes. The problem is that if  `logger.add("some_file.log")` is called at the root level of a Python module, then the file will first be created and owned by the main process. Later, when a child process is spawned by using `multiprocessing` or a web framework such as Guvicorn, it'll re-execute the exact same code and try to open `"some_file.log"` a second time. This causes an error, since two processes aren't allowed to use the same file-object in parallel.

I'll take a look at `filelock` but right now my idea is to provide a new method to ease configuration of the logger in such situation, by using a `socket` to communicate between different processes.
